### DISCUSSION TITLE
Where the users are, December 2021

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#5982] [Show and tell]
January numbers aren't out yet, but our last report was stale...

## By http4s version

| http4s | share | share >= |
| ------ | ----- | -------- |
| 0.1    | 0%    | 100%     |
| 0.2    | 0%    | 100%     |
| 0.3    | 0%    | 100%     |
| 0.4    | 0%    | 100%     |
| 0.5    | 0%    | 100%     |
| 0.6    | 0%    | 100%     |
| 0.7    | 0%    | 100%     |
| 0.8    | 0%    | 100%     |
| 0.9    | 0%    | 100%     |
| 0.10   | 0%    | 100%     |
| 0.11   | 0%    | 100%     |
| 0.12   | 0%    | 100%     |
| 0.13   | 0%    | 100%     |
| 0.14   | 0%    | 99%      |
| 0.15   | 2%    | 99%      |
| 0.16   | 1%    | 97%      |
| 0.17   | 0%    | 96%      |
| 0.18   | 3%    | 96%      |
| 0.19   | 0%    | 93%      |
| 0.20   | 4%    | 93%      |
| 0.21   | 42%   | 89%      |
| 0.22   | 8%    | 48%      |
| 0.23   | 25%   | 40%      |
| 1.0    | 15%   | 15%      |

## By Scala

| scala  | share | share >= |
| ------ | ----- | -------- |
| \_2.10 | 0%    | 100%     |
| \_2.11 | 3%    | 100%     |
| \_2.12 | 35%   | 97%      |
| \_2.13 | 58%   | 62%      |
| \_3    | 4%    | 4%       |

## By Platform

| platform | share |
| -------- | ----- |
| JVM      | 99%   |
| JS       | 1%    |

## compiler, by http4s version

|      | \_2.11 | \_2.12 | \_2.13 | \_3 | \_2.12\_sjs1 | \_2.13\_sjs1 | \_3\_sjs1 |
| ---- | ------ | ------ | ------ | --- | ------------ | ------------ | --------- |
| 0.18 | 16%    | 84%    | 0%     | 0%  | 0%           | 0%           | 0%        |
| 0.20 | 13%    | 87%    | 0%     | 0%  | 0%           | 0%           | 0%        |
| 0.21 | 0%     | 44%    | 56%    | 0%  | 0%           | 0%           | 0%        |
| 0.22 | 0%     | 30%    | 66%    | 4%  | 0%           | 0%           | 0%        |
| 0.23 | 0%     | 24%    | 62%    | 11% | 1%           | 2%           | 1%        |
| 1.0  | 0%     | 7%     | 89%    | 3%  | 0%           | 0%           | 0%        |

## http4s version, by compiler

|              | 0.18 | 0.20 | 0.21 | 0.22 | 0.23 | 1.0 |
| ------------ | ---- | ---- | ---- | ---- | ---- | --- |
| \_2.11       | 48%  | 51%  | 1%   | 0%   | 0%   | 0%  |
| \_2.12       | 7%   | 9%   | 55%  | 7%   | 18%  | 3%  |
| \_2.13       | 0%   | 0%   | 41%  | 9%   | 27%  | 23% |
| \_3          | 0%   | 0%   | 0%   | 9%   | 78%  | 13% |
| \_2.12\_sjs1 | 0%   | 0%   | 0%   | 0%   | 76%  | 24% |
| \_2.13\_sjs1 | 0%   | 0%   | 0%   | 0%   | 90%  | 10% |
| \_3\_sjs1    | 0%   | 0%   | 0%   | 0%   | 85%  | 15% |

**Comment by rossabaker** [comment]
### http4s-0.22

Defense: 60% of our users are still on Cats Effect 2.  Over half our users are on an supported version (:scream:), so upgrade paths should be easy.

Condemnation: it was intended as a Scala 3 upgrade path, and 91% of our Scala 3 users passed it by.

I think it's time to merge the existing PRs on 0.22 and transition to bugfixes only.

**Comment by rossabaker** [comment]
### Scala 2.12

Defense: Milestones draw an unrepresentative sample, and 38% of our users have to go somewhere.

Condemnation: only 7% of 1.0 adopters are on Scala 2.12, which will decline further before 1.0

I think we should proceed with #5337 and if there is demand at RC time, execute @bplommer's plan on #5289.  I think we need to keep it for 0.22 and 0.23.

**Reply by armanbilge** [reply]
We definitely need to keep 2.12 for 0.23 as long as sbt 1.x is a thing. I'm pretty sure 0.23 and 1.0 can never co-exist in the sbt 1.x ecosystem.

**Reply by rossabaker** [reply]
The only alternative I can imagine is to change the base package, but I wouldn't do that just for SBT plugins.

**Comment by armanbilge** [comment]
> JS: 1%

I know this seems impressive, but I'm pretty sure this is all CI 😜

**Reply by cquiroz** [reply]
I bet it is just us in gemini and CI

**Comment by janilcgarcia** [comment]
I don't know if that provides any insights, but I'm starting a new project and I'm on 0.23 because Caliban is locking me (I don't want to write an adapter). I'm using Scala 3.

**Reply by rossabaker** [reply]
I'm not maintaining a 1.x branch for most downstream projects, either.  I imagine once it's stable, more projects will jump on board quickly.
